docs/user-guide/custom_model.md

```diff
@@ -0,0 +1,71 @@
+
+### HTTP_JSON Custom Model
+Users can add the artifact (model or anything else) in addition to the docker image when uploading the model. During the deployment, these artifacts will be made available in the directory specified by `CARAML_ARTIFACT_LOCATION` environment variable.
+
+Web service must open and listen to the port number given by `CARAML_HTTP_PORT` environment variable. 
+
+Web service MUST implement the following endpoints:
+| Endpoint | HTTP Method | Description|
+|--------- |-------------|------------|
+| `/v1/models/{model_name}:predict`| POST | For every inference or prediction calls, it will call this endpoint. Merlin will give the `CARAML_MODEL_FULL_NAME` environment variable, this value can be used as {model_name} for this endpoint. |
+| `/v1/models/{model_name}` | GET | This endpoint will be used to check model healthiness. Model can serve after this API return 200 status code. |
+| `/` | GET | This endpoint will be used as server liveness. Return 200 if the model is healthy. |
+| `/metrics` | GET | This endpoint is used by prometheus to pull the metrics produced by the predictor. The implementation of this endpoint is handled by prometheus library, for example [this](https://prometheus.io/docs/guides/go-application/) is how to implement the endpoint with golang.
+
+### UPI_V1 Custom Model
+Similar with `HTTP_JSON` custom model, user can add the artifact during model uploading, and the uploaded artifacts will be available in the directory specified by `CARAML_ARTIFACT_LOCATION` environment variable. The web server must implement service that defined in the [UPI interface](https://github.com/caraml-dev/universal-prediction-interface/blob/main/proto/caraml/upi/v1/upi.proto#L11), also the server must open and listen to the port number given by `CARAML_GRPC_PORT` environment variable.
+
+If users want to emit metrics from this web server, they need to create scrape metrics REST endpoint. The challenge here, the knative (the underlying k8s deployment tools that merlin use) doesn't open multiple ports, hence the REST endpoint must be running on the same port as gRPC server (using port number given by `CARAML_GRPC_PORT`). Not every programming language can support running multiple protocol (gRPC and HTTP in this case) on the same port, for Go language users can use [cmux](https://github.com/soheilhy/cmux) to solve this problem, otherwise users can use push metrics to [pushgateway](https://prometheus.io/docs/instrumenting/pushing/)
+
+## Docker Image
+
+Docker image must contains web service application and dependencies that must be installed in order to run the web service. Users are responsible for building the docker image as well as for publishing it. Currently there are only two kind of docker registries that can be used:
```

Comment:
We can remove this part since it's Gojek-specific
```
 Currently there are only two kind of docker registries that can be used:
* Artifactory
* GCR that is owned by gods GCP project
```